hail/python/test/hailtop/aiotools/test_copy.py

```diff
@@ -416,3 +417,40 @@ async def test_copy_src_parts(copy_test_context):
 
     await expect_file(fs, f'{dest_base}file1', 'src/a/file1')
     await expect_file(fs, f'{dest_base}subdir/file2', 'src/a/subdir/file2')
+
+
+@pytest.mark.asyncio
+async def test_file_and_directory_error_with_slash(router_filesystem):
+    sema, fs, bases = router_filesystem
+
+    src_base = await fresh_dir(fs, bases, 'gs')
+    dest_base = await fresh_dir(fs, bases, 'gs')
+
+    # test empty file ending in slash and directory
+    await fs.create(f'{src_base}empty/')
+    async with await fs.create(f'{src_base}empty/foo') as f:
+        await f.write(b'foo')
+
+    await fs.copy(sema, Transfer(f'{src_base}empty/', dest_base.rstrip('/')))
+    await expect_file(fs, f'{dest_base}empty/foo', 'foo')
+    with pytest.raises(AssertionError):
+        await expect_file(fs, f'{dest_base}empty/', '')
+
+    # Test non-empty file ending in slash and directory
+    async with await fs.create(f'{src_base}not-empty/') as f:
+        await f.write(b'not-empty')
+
+    async with await fs.create(f'{src_base}not-empty/bar') as f:
+        await f.write(b'bar')
+
+    await fs.copy(sema, Transfer(f'{src_base}not-empty/bar', dest_base.rstrip('/')))
+    await expect_file(fs, f'{dest_base}not-empty/bar', 'bar')
+
+    with pytest.raises(FileAndDirectoryError):
+        await fs.copy(sema, Transfer(f'{src_base}not-empty/', dest_base.rstrip('/')))
+
+    # Test empty file that ends in a slash
+    await fs.create(f'{src_base}empty-only/')
+
+    with pytest.raises(FileNotFoundError):
```

Comment:
I think this shouldn't be FileNotFound.  {src_base}empty_only/ exists and is a directory.  This should work, see my cp example in another comment.